README.rst

====
pgqd
====

--------------------------
Maintenance daemon for PgQ
--------------------------

:Manual section: 1

Synopsis
========

pgqd [-qvd] config

pgqd [-skr] config

pgqd --ini|-h|-V

Description
===========

Runs both ticker and periodic maintenence for all
databases in one PostgreSQL cluster.

Options
=======

-q      Do not log to stdout
-v      Verbose log
-d      Daemonize process
-s      Send SIGINT to running process to stop it
-k      Send SIGTERM to running process to stop it
-r      Send SIGHUP to running process to reload config
-h      Show help
-V      Show version
--ini   Show sample config

Configuration
=============

Config uses `ini` file syntax::

    [pgqd]
    logfile = ~/log/pgqd.log
    pidfile = ~/pid/pgqd.pid

Options:

logfile
    Filename to log to.
    Default: empty.

pidfile
    Filename to store pid, required when daemonizing.
    Default: empty.

base_connstr
    Connect string without dbname=
    Default: empty.

initial_database
    Startup db to query other databases.
    Default: template1

database_list
    Limit ticker to specific databases.
    Default: empty, which means all database.

syslog
    Whether to log into syslog.
    Default: 1

syslog_ident
    Name to use for syslog.
    Default: pgqd

check_period
    How often to check for new databases, in seconds.
    Default: 60.

retry_period
    How often to flush retry queue, in seconds.
    Default: 30

maint_period
    How often to do maintentance, in seconds.
    Default: 120

ticker_period
    How often to run ticker, in seconds.
    Default: 1


Install
=======

pgqd uses autoconf based build system::

    ./configure --prefix=/opt
    make
    make install

Dependencies: libevent, python3-docutils



NEWS.rst
NEWS
====

pgqd 3.5
--------

* Quote dbname in logs.
* Upgrade libusual.
* Add CI workflows.

pgqd 3.4
--------

* Upgrade libusual.
* Switch to direct libevent2 usage.
* Fix ``make dist``.
* Add manpage.



tests/test.sh
#! /usr/bin/env bash

cd $(dirname $0)

PGQD=../pgqd

export PGHOST="${PGHOST:-127.0.0.1}"
export PGPORT="${PGPORT:-5432}"
export LANG=C
export LC_ALL=C

LOGDIR=log

set -o pipefail

mkdir -p log
rm -f log/*

ulimit -c unlimited

echo "Creating databases"
for dbname in db1; do
    dropdb --if-exists $dbname
    createdb $dbname || exit 1
    psql -q -d $dbname -c "create extension pgq"
done

die() {
    echo $@
    exit 1
}

runtest() {
    local status

    printf "`date` running $1 ... "
    conf="${LOGDIR}/$1.ini"
    logfile="${LOGDIR}/$1.log"
    outfile="${LOGDIR}/$1.out"
    pidfile="${LOGDIR}/$1.pid"
    printf "[pgqd]\nlogfile=${logfile}\npidfile=${pidfile}\ncheck_period=3\nmaint_period=7\n" > "${conf}"
    eval "$1" "${conf}" "${logfile}" > ${outfile} 2>&1
    res=$?
    date >> ${outfile}
    if [ $res -eq 0 ]; then
        echo "ok"
    else
        echo "FAILED"
        cat ${outfile} | sed 's/^/out> /'
        if test -f "${logfile}"; then
            cat ${logfile} | sed 's/^/log> /'
        fi
    fi

    test -f "${pidfile}" && kill $(cat "${pidfile}")

    return $res
}

#
# testcases
#

test_version() {
    ln=$(${PGQD} -V)
    case "$ln" in
        *version*) res=0;;
        *) res=1;;
    esac
    return $res
}

test_show_ini() {
    ${PGQD} --ini | grep -q 'logfile'
    return $?
}

test_stop() {
    ${PGQD} -d $1 || return 1
    sleep 1
    ${PGQD} -s $1 || return 1
    sleep 1
    grep -q SIGINT $2
    return $?
}

test_reload() {
    ${PGQD} -d $1 || return 1
    sleep 1
    ${PGQD} -r $1 || return 1
    sleep 1
    grep -q SIGHUP $2
    res=$?
    ${PGQD} -s $1 || return 1
    sleep 1
    return $res
}

create_queue() {
    db="$1"
    qname="$2"
    psql -q -d "${db}" -c "select pgq.create_queue('${qname}')" \
        || die "queue creation failed"
    psql -q -d "${db}" -c "update pgq.queue set queue_rotation_period='10 seconds' where queue_name='${qname}'" \
        || die "queue setup failed"
}

test_ticker() {
    db="db1"
    qname="test1"
    create_queue "${db}" "${qname}"
    ${PGQD} -d "$1" || return 1
    sleep 3
    tick_id=$(psql -q -d "${db}" -At -c "select last_tick_id from pgq.get_queue_info('${qname}')")
    echo "tick_id=${tick_id}"
    test "${tick_id}" -gt 2 || return 1
    return 0
}

test_maint() {
    db="db1"
    qname="test2"
    create_queue "${db}" "${qname}"
    ${PGQD} -d "$1" || return 1
    sleep 15
    cur_tbl=$(psql -q -d "${db}" -At -c "select queue_cur_table from pgq.get_queue_info('${qname}')")
    echo "cur_tbl=${cur_tbl}"
    test "${cur_tbl}" -gt 0 || return 1
    return 0
}

testlist="
test_version
test_show_ini
test_stop
test_reload
test_ticker
test_maint
"

if [ $# -gt 0 ]; then
    testlist="$*"
fi

final=0
for testcase in $testlist; do
    runtest $testcase || final=1
done

exit $final



src/pgqd.h

#ifndef __PGQD_H__
#define __PGQD_H__

#include <string.h>

#define Assert(x)

#include <usual/list.h>
#include <usual/statlist.h>
#include <usual/logging.h>

#include "pgsocket.h"

enum DbState {
	DB_CLOSED,
	DB_TICKER_CHECK_PGQ,
	DB_TICKER_CHECK_VERSION,
	DB_TICKER_RUN,
	DB_MAINT_TEST_VERSION,
	DB_MAINT_LOAD_OPS,
	DB_MAINT_OP,
	DB_MAINT_LOAD_QUEUES,
	DB_MAINT_ROT1,
	DB_MAINT_ROT2,
	DB_MAINT_VACUUM_LIST,
	DB_MAINT_DO_VACUUM,
};

struct MaintOp;

struct PgDatabase {
	struct List head;
	const char *name;
	const char *logname;
	struct PgSocket *c_ticker;
	struct PgSocket *c_maint;
	struct PgSocket *c_retry;
	bool has_pgq;
	enum DbState state;
	enum DbState maint_state;
	bool dropped;

	struct StrList *maint_item_list;
	struct StatList maint_op_list;
	struct MaintOp *cur_maint;

	bool has_maint_operations;
};

struct Config {
	const char *config_file;
	const char *pidfile;
	const char *base_connstr;
	const char *initial_database;
	const char *database_list;

	double retry_period;
	double check_period;
	double maint_period;
	double ticker_period;
	double stats_period;

	double connection_lifetime;
};

struct Stats {
	int n_ticks;
	int n_maint;
	int n_retry;
};

extern struct Config cf;
extern struct Stats stats;
extern struct event_base *ev_base;

void launch_ticker(struct PgDatabase *db);
void launch_maint(struct PgDatabase *db);
void launch_retry(struct PgDatabase *db);

void free_maint(struct PgDatabase *db);

char *make_connstr(const char *dbname);

#endif



src/maint.c
#include "pgqd.h"

#include <usual/string.h>
#include <usual/pgutil.h>

#include <ctype.h>

struct MaintOp {
	struct List head;
	const char *func_name;
	const char *func_arg;
};

static struct MaintOp *next_op(struct PgDatabase *db)
{
	struct List *el = statlist_pop(&db->maint_op_list);
	if (!el)
		return NULL;
	return container_of(el, struct MaintOp, head);
}

static void free_op(struct MaintOp *op)
{
	if (op) {
		free((void *)op->func_name);
		free((void *)op->func_arg);
		free(op);
	}
}

void free_maint(struct PgDatabase *db)
{
	struct MaintOp *op;

	strlist_free(db->maint_item_list);
	db->maint_item_list = NULL;

	while ((op = next_op(db)) != NULL) {
		free_op(op);
	}
	free_op(db->cur_maint);
	db->cur_maint = NULL;
}

static void close_maint(struct PgDatabase *db, double sleep_time)
{
	log_debug("%s: close_maint, %f", db->logname, sleep_time);
	db->maint_state = DB_CLOSED;
	pgs_reconnect(db->c_maint, sleep_time);
}

static void run_test_version(struct PgDatabase *db)
{
	const char *q = "select 1 from pg_proc p, pg_namespace n"
			" where p.pronamespace = n.oid"
			"   and p.proname = 'maint_operations'"
			"   and n.nspname = 'pgq'";
	log_debug("%s: %s", db->logname, q);
	pgs_send_query_simple(db->c_maint, q);
	db->maint_state = DB_MAINT_TEST_VERSION;
}

static bool has_ops(PGresult *res)
{
	if (PQntuples(res) == 1 && atoi(PQgetvalue(res, 0, 0)) == 1)
		return true;
	return false;
}

static bool fill_op_list(struct PgDatabase *db, PGresult *res)
{
	int i;
	struct MaintOp *op = NULL;
	const char *fname, *farg;

	free_maint(db);

	for (i = 0; i < PQntuples(res); i++) {
		op = calloc(1, sizeof(*op));
		if (!op)
			return false;
		list_init(&op->head);
		fname = PQgetvalue(res, i, 0);
		farg = NULL;
		if (!PQgetisnull(res, i, 1))
			farg = PQgetvalue(res, i, 1);
		log_debug("load_op: %s / %s", fname, farg ? farg : "NULL");
		op->func_name = strdup(fname);
		if (!op->func_name)
			goto failed;
		if (farg) {
			op->func_arg = strdup(farg);
			if (!op->func_arg)
				goto failed;
		}
		statlist_append(&db->maint_op_list, &op->head);
	}
	return true;
failed:
	free_op(op);
	return false;
}

static void run_op_list(struct PgDatabase *db)
{
	const char *q = "select func_name, func_arg from pgq.maint_operations()";
	log_debug("%s: %s", db->logname, q);
	pgs_send_query_simple(db->c_maint, q);
	db->maint_state = DB_MAINT_LOAD_OPS;
}

static const char *stmt_names[] = {
	"vacuum",
	"vacuum analyze",
	NULL
};

static void run_op(struct PgDatabase *db, PGresult *res)
{
	struct MaintOp *op;
	char buf[1024];
	char namebuf[256];
	const char **np;

	if (db->cur_maint) {
		if (res && PQntuples(res) > 0) {
			const char *val = PQgetvalue(res, 0, 0);
			if (val && atoi(val)) {
				op = db->cur_maint;
				goto repeat;
			}
		}
next:
		free_op(db->cur_maint);
		db->cur_maint = NULL;
	}
	op = next_op(db);
	if (!op) {
		stats.n_maint++;
		close_maint(db, cf.maint_period);
		return;
	}
	db->cur_maint = op;
repeat:
	/* check if its magic statement */
	for (np = stmt_names; *np; np++) {
		if (strcasecmp(op->func_name, *np) != 0)
			continue;
		if (!pg_quote_fqident(namebuf, op->func_arg, sizeof(namebuf))) {
			log_error("Bad table name? - %s", op->func_arg);
			goto next;
		}
		/* run as a statement */
		snprintf(buf, sizeof(buf), "%s %s", op->func_name, namebuf);
		log_debug("%s: [%s]", db->logname, buf);
		pgs_send_query_simple(db->c_maint, buf);
		goto done;
	}

	/* run as a function */
	if (!pg_quote_fqident(namebuf, op->func_name, sizeof(namebuf))) {
		log_error("Bad func name? - %s", op->func_name);
		goto next;
	}
	if (op->func_arg) {
		snprintf(buf, sizeof(buf), "select %s($1)", namebuf);
		log_debug("%s: [%s]", db->logname, buf);
		pgs_send_query_params(db->c_maint, buf, 1, op->func_arg);
	} else {
		snprintf(buf, sizeof(buf), "select %s()", namebuf);
		log_debug("%s: [%s]", db->logname, buf);
		pgs_send_query_simple(db->c_maint, buf);
	}
done:
	db->maint_state = DB_MAINT_OP;
}

static bool fill_items(struct PgDatabase *db, PGresult *res)
{
	int i;
	if (db->maint_item_list)
		strlist_free(db->maint_item_list);
	db->maint_item_list = strlist_new(USUAL_ALLOC);
	if (!db->maint_item_list)
		return false;
	for (i = 0; i < PQntuples(res); i++) {
		const char *item = PQgetvalue(res, i, 0);
		if (item)
			if (!strlist_append(db->maint_item_list, item))
				return false;
	}
	return true;
}

static void run_queue_list(struct PgDatabase *db)
{
	const char *q = "select queue_name from pgq.get_queue_info()";
	log_debug("%s: %s", db->logname, q);
	pgs_send_query_simple(db->c_maint, q);
	db->maint_state = DB_MAINT_LOAD_QUEUES;
}

static void run_vacuum_list(struct PgDatabase *db)
{
	const char *q = "select * from pgq.maint_tables_to_vacuum()";
	log_debug("%s: %s", db->logname, q);
	pgs_send_query_simple(db->c_maint, q);
	db->maint_state = DB_MAINT_VACUUM_LIST;
}

static void run_rotate_part1(struct PgDatabase *db)
{
	const char *q;
	const char *qname;
	qname = strlist_pop(db->maint_item_list);
	q = "select pgq.maint_rotate_tables_step1($1)";
	log_debug("%s: %s [%s]", db->logname, q, qname);
	pgs_send_query_params(db->c_maint, q, 1, qname);
	free((void *)qname);
	db->maint_state = DB_MAINT_ROT1;
}

static void run_rotate_part2(struct PgDatabase *db)
{
	const char *q = "select pgq.maint_rotate_tables_step2()";
	log_debug("%s: %s", db->logname, q);
	pgs_send_query_simple(db->c_maint, q);
	db->maint_state = DB_MAINT_ROT2;
}

static void run_vacuum(struct PgDatabase *db)
{
	char qbuf[256];
	const char *table;
	table = strlist_pop(db->maint_item_list);
	snprintf(qbuf, sizeof(qbuf), "vacuum %s", table);
	log_debug("%s: %s", db->logname, qbuf);
	pgs_send_query_simple(db->c_maint, qbuf);
	free((void *)table);
	db->maint_state = DB_MAINT_DO_VACUUM;
}

static void maint_handler(struct PgSocket *s, void *arg, enum PgEvent ev, PGresult *res)
{
	struct PgDatabase *db = arg;

	switch (ev) {
	case PGS_CONNECT_OK:
		log_debug("%s: starting maintenance", db->logname);
		if (db->has_maint_operations)
			run_op_list(db);
		else
			run_test_version(db);
		break;
	case PGS_RESULT_OK:
		if (PQresultStatus(res) != PGRES_TUPLES_OK) {
			close_maint(db, 20);
			return;
		}
		switch (db->maint_state) {
		case DB_MAINT_TEST_VERSION:
			if (has_ops(res)) {
				db->has_maint_operations = true;
				run_op_list(db);
			} else {
				run_queue_list(db);
			}
			break;
		case DB_MAINT_LOAD_OPS:
			if (!fill_op_list(db, res))
				goto mem_err;
			/* fallthrough */
		case DB_MAINT_OP:
			run_op(db, res);
			break;
		case DB_MAINT_LOAD_QUEUES:
			if (!fill_items(db, res))
				goto mem_err;
			/* fallthrough */
		case DB_MAINT_ROT1:
			if (!strlist_empty(db->maint_item_list)) {
				run_rotate_part1(db);
			} else {
				run_rotate_part2(db);
			}
			break;
		case DB_MAINT_ROT2:
			run_vacuum_list(db);
			break;
		case DB_MAINT_VACUUM_LIST:
			if (!fill_items(db, res))
				goto mem_err;
			/* fallthrough */
		case DB_MAINT_DO_VACUUM:
			if (!strlist_empty(db->maint_item_list)) {
				run_vacuum(db);
			} else {
				close_maint(db, cf.maint_period);
			}
			break;
		default:
			fatal("bad state");
			break;
		}
		break;
	case PGS_TIMEOUT:
		log_debug("%s: maint timeout", db->logname);
		if (!pgs_connection_valid(db->c_maint))
			launch_maint(db);
		else
			run_queue_list(db);
		break;
	default:
		log_warning("%s: default reconnect", db->logname);
		pgs_reconnect(db->c_maint, 60);
		break;
	}
	return;
mem_err:
	if (db->maint_item_list) {
		strlist_free(db->maint_item_list);
		db->maint_item_list = NULL;
	}
	pgs_disconnect(db->c_maint);
	pgs_sleep(db->c_maint, 20);
}

void launch_maint(struct PgDatabase *db)
{
	char *cstr;

	log_debug("%s: launch_maint", db->logname);

	if (!db->c_maint) {
		if (db->maint_item_list) {
			strlist_free(db->maint_item_list);
			db->maint_item_list = NULL;
		}
		cstr = make_connstr(db->name);
		if (!cstr) {
			log_error("make_connstr: %s", strerror(errno));
			return;
		}
		db->c_maint = pgs_create(cstr, maint_handler, db, ev_base);
		free(cstr);
		if (!db->c_maint) {
			log_error("pgs_create: %s", strerror(errno));
			return;
		}
	}

	if (!pgs_connection_valid(db->c_maint)) {
		pgs_connect(db->c_maint);
	} else {
		/* Already have a connection, what are we doing here */
		log_error("%s: maint already initialized", db->logname);
		return;
	}
}



src/retry.c

#include "pgqd.h"


static void close_retry(struct PgDatabase *db, double sleep_time)
{
	log_debug("%s: close_retry, %f", db->logname, sleep_time);
	pgs_reconnect(db->c_retry, sleep_time);
}

static void run_retry(struct PgDatabase *db)
{
	const char *q = "select * from pgq.maint_retry_events()";
	log_debug("%s: %s", db->logname, q);
	pgs_send_query_simple(db->c_retry, q);
}

static void parse_retry(struct PgDatabase *db, PGresult *res)
{
	if (PQntuples(res) == 1) {
		char *val = PQgetvalue(res, 0, 0);
		stats.n_retry += atoi(val);
		if (strcmp(val, "0") != 0) {
			run_retry(db);
			return;
		}
	}
	close_retry(db, cf.retry_period);
}

static void retry_handler(struct PgSocket *s, void *arg, enum PgEvent ev, PGresult *res)
{
	struct PgDatabase *db = arg;

	switch (ev) {
	case PGS_CONNECT_OK:
		log_debug("%s: starting retry event processing", db->logname);
		run_retry(db);
		break;
	case PGS_RESULT_OK:
		if (PQresultStatus(res) != PGRES_TUPLES_OK)
			close_retry(db, 20);
		else
			parse_retry(db, res);
		break;
	case PGS_TIMEOUT:
		log_debug("%s: retry timeout", db->logname);
		launch_retry(db);
		break;
	default:
		log_warning("%s: default reconnect", db->logname);
		pgs_reconnect(db->c_retry, 30);
		break;
	}
}

void launch_retry(struct PgDatabase *db)
{
	char *cstr;
	if (db->c_retry) {
		log_debug("%s: retry already initialized", db->logname);
	} else {
		log_debug("%s: launch_retry", db->logname);
		cstr = make_connstr(db->name);
		if (!cstr) {
			log_error("make_connstr: %s", strerror(errno));
			return;
		}
		db->c_retry = pgs_create(cstr, retry_handler, db, ev_base);
		free(cstr);
		if (!db->c_retry) {
			log_error("pgs_create: %s", strerror(errno));
			return;
		}
	}
	pgs_connect(db->c_retry);
}



src/pgqd.c
#include "pgqd.h"

#include <getopt.h>

#include <event2/event.h>

#include <usual/string.h>
#include <usual/signal.h>
#include <usual/daemon.h>
#include <usual/cfparser.h>
#include <usual/time.h>

static void detect_dbs(void);
static void recheck_dbs(void);

static const char usage_str[] =
"usage: pgq-ticker [switches] config.file\n"
"Switches:\n"
"  -v        Increase verbosity\n"
"  -q        No output to console\n"
"  -d        Daemonize\n"
"  -h        Show help\n"
"  -V        Show version\n"
" --ini      Show sample config file\n"
"  -s        Stop - send SIGINT to running process\n"
"  -k        Kill - send SIGTERM to running process\n"
#ifdef SIGHUP
"  -r        Reload - send SIGHUP to running process\n"
#endif
"";

static const char *sample_ini =
#include "pgqd.ini.h"
;

struct Config cf;

struct Stats stats;

static struct PgSocket *db_template;

static STATLIST(database_list);

static int got_sigint;

static struct event *ev_stats;
static struct event *ev_sigterm;
static struct event *ev_sigint;
#ifdef SIGHUP
static struct event *ev_sighup;
#endif

struct event_base *ev_base;

#define CF_REL_BASE struct Config
static const struct CfKey conf_params[] = {
	CF_ABS("logfile", CF_FILE, cf_logfile, 0, NULL),
	CF_REL("pidfile", CF_FILE, pidfile, 0, NULL),
	CF_REL("initial_database", CF_STR, initial_database, 0, "template1"),
	CF_REL("base_connstr", CF_STR, base_connstr, 0, ""),
	CF_REL("database_list", CF_STR, database_list, 0, NULL),
	CF_ABS("syslog", CF_INT, cf_syslog, 0, "1"),
	CF_ABS("syslog_ident", CF_STR, cf_syslog_ident, 0, "pgqd"),
	CF_ABS("syslog_facility", CF_STR, cf_syslog_facility, 0, "daemon"),
	CF_REL("check_period", CF_TIME_DOUBLE, check_period, 0, "60"),
	CF_REL("maint_period", CF_TIME_DOUBLE, maint_period, 0, "120"),
	CF_REL("retry_period", CF_TIME_DOUBLE, retry_period, 0, "30"),
	CF_REL("ticker_period", CF_TIME_DOUBLE, ticker_period, 0, "1"),
	CF_REL("stats_period", CF_TIME_DOUBLE, stats_period, 0, "30"),
	CF_REL("connection_lifetime", CF_TIME_DOUBLE, connection_lifetime, 0, "3600"),
	{ NULL },
};

static const struct CfSect conf_sects[] = {
	{ "pgqd", conf_params },
	{ NULL }
};

static struct CfContext conf_info = {
	.sect_list = conf_sects,
	.base = &cf,
};

static void load_config(void)
{
	bool ok = cf_load_file(&conf_info, cf.config_file);
	if (!ok)
		fatal("failed to read config");
	reset_logging();
}

static void handle_sigterm(evutil_socket_t sock, short flags, void *arg)
{
	log_info("Got SIGTERM, fast exit");
	/* pidfile cleanup happens via atexit() */
	exit(1);
}

static void handle_sigint(evutil_socket_t sock, short flags, void *arg)
{
	log_info("Got SIGINT, shutting down");
	/* notify main loop to exit */
	got_sigint = 1;
}

#ifdef SIGHUP
static void handle_sighup(evutil_socket_t sock, short flags, void *arg)
{
	log_info("Got SIGHUP, re-reading config");
	load_config();
	recheck_dbs();
}
#endif

static void signal_setup(void)
{
	int err;

#ifdef SIGPIPE
	sigset_t set;

	/* block SIGPIPE */
	sigemptyset(&set);
	sigaddset(&set, SIGPIPE);
	err = sigprocmask(SIG_BLOCK, &set, NULL);
	if (err < 0)
		fatal_perror("sigprocmask");
#endif

#ifdef SIGHUP
	/* catch signals */
	ev_sighup = evsignal_new(ev_base, SIGHUP, handle_sighup, NULL);
	if (!ev_sighup)
		fatal_perror("evsignal_new");
	err = evsignal_add(ev_sighup, NULL);
	if (err < 0)
		fatal_perror("evsignal_add");
#endif

	ev_sigterm = evsignal_new(ev_base, SIGTERM, handle_sigterm, NULL);
	if (!ev_sigterm)
		fatal_perror("evsignal_new");
	err = evsignal_add(ev_sigterm, NULL);
	if (err < 0)
		fatal_perror("evsignal_add");

	ev_sigint = evsignal_new(ev_base, SIGINT, handle_sigint, NULL);
	if (!ev_sigint)
		fatal_perror("evsignal_new");
	err = evsignal_add(ev_sigint, NULL);
	if (err < 0)
		fatal_perror("signal_add");
}

char *make_connstr(const char *dbname)
{
	size_t buflen;
	char *buf, *dst;
	const char *src;

	buflen = strlen(cf.base_connstr) + strlen(dbname) * 2 + 32;
	buf = calloc(1, buflen);
	if (!buf)
		return NULL;
	snprintf(buf, buflen, "%s dbname='", cf.base_connstr);
	dst = buf + strlen(buf);
	for (src = dbname; *src; src++) {
		if (*src == '\'' || *src == '\\') {
			*dst++ = '\\';
		}
		*dst++ = *src;
	}
	*dst = '\'';
	return buf;
}

static char *safe_dbname(const char *dbname)
{
	char *buf, *dst;
	const char *src;
	size_t buflen;

	buflen = strlen(dbname) * 2 + 3;
	buf = calloc(1, buflen);
	if (!buf)
		return NULL;
	dst = buf;
	*dst++ = '[';
	for (src = dbname; *src; src++) {
		if ((unsigned char)*src < ' ') {
			*dst++ = '?';
		} else {
			*dst++ = *src;
		}
	}
	*dst++ = ']';
	*dst++ = '\0';
	return buf;
}

static void launch_db(const char *dbname)
{
	struct PgDatabase *db;
	struct List *elem;

	/* check of already exists */
	statlist_for_each(elem, &database_list) {
		db = container_of(elem, struct PgDatabase, head);
		if (strcmp(db->name, dbname) == 0) {
			db->dropped = false;
			return;
		}
	}

	/* create new db entry */
	db = calloc(1, sizeof(*db));
	if (!db) {
		log_error("calloc: %s", strerror(errno));
		return;
	}
	db->name = strdup(dbname);
	if (!db->name) {
		log_error("strdup: %s", strerror(errno));
		free(db);
		return;
	}
	db->logname = safe_dbname(dbname);
	if (!db->logname) {
		log_error("safe_dbname: %s", strerror(errno));
		free((void *)db->name);
		free(db);
		return;
	}
	list_init(&db->head);
	statlist_init(&db->maint_op_list, "maint_op_list");
	statlist_append(&database_list, &db->head);

	/* start working on it */
	launch_ticker(db);
}

static void drop_db(struct PgDatabase *db, bool log)
{
	if (log)
		log_info("Unregister database: %s", db->logname);
	statlist_remove(&database_list, &db->head);
	pgs_free(db->c_ticker);
	pgs_free(db->c_maint);
	pgs_free(db->c_retry);
	free_maint(db);
	free((void *)db->logname);
	free((void *)db->name);
	free(db);
}

static void detect_handler(struct PgSocket *sk, void *arg, enum PgEvent ev, PGresult *res)
{
	int i;
	const char *s;
	struct List *el, *tmp;
	struct PgDatabase *db;

	switch (ev) {
	case PGS_CONNECT_OK:
		pgs_send_query_simple(sk, "select datname from pg_database"
				     	 " where not datistemplate and datallowconn");
		break;
	case PGS_RESULT_OK:
		/* tag old dbs as dead */
		statlist_for_each(el, &database_list) {
			db = container_of(el, struct PgDatabase, head);
			db->dropped = true;
		}
		/* process new dbs */
		for (i = 0; i < PQntuples(res); i++) {
			s = PQgetvalue(res, i, 0);
			launch_db(s);
		}
		/* drop old dbs */
		statlist_for_each_safe(el, &database_list, tmp) {
			db = container_of(el, struct PgDatabase, head);
			if (db->dropped)
				drop_db(db, true);
		}
		pgs_disconnect(sk);
		pgs_sleep(sk, cf.check_period);
		break;
	case PGS_TIMEOUT:
		detect_dbs();
		break;
	default:
		pgs_disconnect(sk);
		pgs_sleep(sk, cf.check_period);
		break;
	}
}

static void detect_dbs(void)
{
	if (!db_template) {
		char *cstr = make_connstr(cf.initial_database);
		if (!cstr) {
			log_error("make_connstr: %s", strerror(errno));
			return;
		}
		db_template = pgs_create(cstr, detect_handler, NULL, ev_base);
		free(cstr);
		if (!db_template) {
			log_error("pgs_create: %s", strerror(errno));
			return;
		}
	}
	pgs_connect(db_template);
}

static bool launch_db_cb(void *arg, const char *db)
{
	launch_db(db);
	return true;
}

static void recheck_dbs(void)
{
	struct PgDatabase *db;
	struct List *el, *tmp;
	if (cf.database_list && cf.database_list[0]) {
		/* tag old dbs as dead */
		statlist_for_each(el, &database_list) {
			db = container_of(el, struct PgDatabase, head);
			db->dropped = true;
		}
		/* process new ones */
		if (!parse_word_list(cf.database_list, launch_db_cb, NULL)) {
			log_warning("database_list parsing failed: %s", strerror(errno));
			return;
		}
		/* drop old ones */
		statlist_for_each_safe(el, &database_list, tmp) {
			db = container_of(el, struct PgDatabase, head);
			if (db->dropped)
				drop_db(db, true);
		}

		/* done with template for the moment */
		if (db_template) {
			pgs_free(db_template);
			db_template = NULL;
		}
	} else if (!db_template) {
		log_info("auto-detecting dbs ...");
		detect_dbs();
	}
}


static void stats_handler(evutil_socket_t fd, short flags, void *arg)
{
	struct timeval tv = { cf.stats_period, 0 };

	log_info("{ticks: %d, maint: %d, retry: %d}",
		 stats.n_ticks, stats.n_maint, stats.n_retry);
	memset(&stats, 0, sizeof(stats));

	if (evtimer_add(ev_stats, &tv) < 0)
		fatal_perror("evtimer_add");
}

static void stats_setup(void)
{
	struct timeval tv = { cf.stats_period, 0 };
	ev_stats = evtimer_new(ev_base, stats_handler, NULL);
	if (!ev_stats)
		fatal_perror("evtimer_new");
	if (evtimer_add(ev_stats, &tv) < 0)
		fatal_perror("evtimer_add");
}

static void cleanup(void)
{
	struct PgDatabase *db;
	struct List *elem, *tmp;

	statlist_for_each_safe(elem, &database_list, tmp) {
		db = container_of(elem, struct PgDatabase, head);
		drop_db(db, false);
	}
	pgs_free(db_template);

	event_base_free(NULL);
	reset_logging();

#ifdef SIGHUP
	event_free(ev_sighup);
#endif
	event_free(ev_sigint);
	event_free(ev_sigterm);
	event_free(ev_stats);
}

static void main_loop_once(void)
{
	reset_time_cache();
	if (event_base_loop(ev_base, EVLOOP_ONCE) != 0) {
		log_error("event_loop failed: %s", strerror(errno));
	}
}

int main(int argc, char *argv[])
{
	int c;
	bool daemon = false;
	int sig = 0;
	const char *signame = NULL;

	for (c = 1; c < argc; c++) {
		if (!strcmp(argv[c], "--ini")) {
			printf("%s", sample_ini);
			exit(0);
		}
		if (!strcmp(argv[c], "--help")) {
			printf(usage_str);
			exit(0);
		}
	}

	while ((c = getopt(argc, argv, "dqvhVrsk")) != -1) {
		switch (c) {
		case 'd':
			daemon = true;
			break;
		case 'v':
			cf_verbose++;
			break;
		case 'q':
			cf_quiet = 1;
			break;
		case 'h':
			printf(usage_str);
			return 0;
		case 'V':
			printf("%s version %s\n", PACKAGE_NAME, PACKAGE_VERSION);
			return 0;
#ifdef SIGHUP
		case 'r':
			sig = SIGHUP;
			signame = "SIGHUP";
			break;
#endif
		case 's':
			sig = SIGINT;
			signame = "SIGINT";
			break;
		case 'k':
			sig = SIGTERM;
			signame = "SIGTERM";
			break;
		default:
			printf("bad switch: ");
			printf(usage_str);
			return 1;
		}
	}
	if (optind + 1 != argc) {
		fprintf(stderr, "pgqd requires config file\n");
		return 1;
	}

	cf.config_file = argv[optind];

	load_config();
	conf_info.loaded = true;

	if (sig) {
		if (!cf.pidfile || !cf.pidfile[0]) {
			fprintf(stderr, "No pidfile configured\n");
			return 1;
		}
		if (signal_pidfile(cf.pidfile, sig))
			fprintf(stderr, "%s sent\n", signame);
		else
			fprintf(stderr, "Old process is not running\n");
		return 0;
	}

	log_info("Starting pgqd " PACKAGE_VERSION);

	daemonize(cf.pidfile, daemon);

	ev_base = event_base_new();
	if (!ev_base)
		fatal("event_base_new failed");

	signal_setup();

	stats_setup();

	recheck_dbs();

	while (!got_sigint)
		main_loop_once();

	cleanup();
	event_base_free(ev_base);

	return 0;
}
